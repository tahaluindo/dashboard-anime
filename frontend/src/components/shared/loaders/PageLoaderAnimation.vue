<template>
  <div
    class="
      absolute
      inset-0
      z-0
      bg-transparent
      flex flex-col
      justify-center
      items-center
    "
    v-show="show"
  >
    <span ref="text"><slot /></span>
    <span ref="circle" class="w-3 h-3 rounded-full bg-primary"></span>
  </div>
</template>

<script lang="ts">
import { defineComponent, onMounted, ref } from "vue";
import { useRouter } from "vue-router";
import { PageLoaderAnime } from "./controllers/pageLoaderAnime";
import { TextLoaderAnime } from "./controllers/textLoaderAnime";

export default defineComponent({
  setup() {
    const show = ref(false);
    const text = ref<HTMLElement>();
    const circle = ref<HTMLElement>();
    let textAnimation: TextLoaderAnime;
    let circleAnimation: PageLoaderAnime;

    onMounted(() => {
      if (circle.value) circleAnimation = new PageLoaderAnime(circle.value);

      if (text.value && text.value?.firstElementChild) {
        textAnimation = new TextLoaderAnime(
          TextLoaderAnime.prepareLettersToAnimate(
            text.value.firstElementChild as HTMLElement
          )
        );
      }
    });

    const animate = () => {
      show.value = true;
      textAnimation?.animate();
      circleAnimation?.animate();
    };

    const stop = () => {
      setTimeout(() => {
        textAnimation?.stop();
        circleAnimation?.stop();
        show.value = false;
      }, 1500);
    };

    const router = useRouter();
    router.beforeEach(() => {
      animate();
      stop();
    });

    return {
      animate,
      stop,
      text,
      circle,
      show,
    };
  },
});
</script>

<style lang="postcss"></style>
