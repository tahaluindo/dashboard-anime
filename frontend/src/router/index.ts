import { createRouter, createWebHistory, RouteRecordRaw } from "vue-router";
import AddUser from "../views/AddUser.vue";
import AllUsers from "../views/AllUsers.vue";
import Home from "../views/Home.vue";
import User from "../views/User.vue";

const routes: Array<RouteRecordRaw> = [
  {
    path: "/",
    name: "Home",
    component: Home,
  },
  {
    path: "/usuarios",
    name: "AllUsers",
    component: AllUsers,
  },
  {
    path: "/usuario/:id",
    component: User,
    name: "User",
    props: true,
  },
  {
    path: "/adicionar-usuario",
    name: "AddUser",
    component: AddUser,
  },
];

const router = createRouter({
  history: createWebHistory(import.meta.env.BASE_URL),
  routes,
  scrollBehavior() {
    return window.scrollTo({ top: 0, behavior: "smooth" });
  },
});

export default router;
