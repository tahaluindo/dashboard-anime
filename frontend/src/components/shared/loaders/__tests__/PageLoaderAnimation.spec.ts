import { flushPromises, mount } from "@vue/test-utils";
import { nextTick } from "vue";
import router from "../../../../router/__mocks__/router.mock";
import PageLoaderAnimation from "../PageLoaderAnimation.vue";

describe("PageLoaderAnimation test suit", () => {
  it("renders the default componenent", async () => {
    jest.useFakeTimers();
    router.push("/");
    await router.isReady();

    const beforeEachSpy = jest.spyOn(router, "beforeEach");

    const wrapper = mount(PageLoaderAnimation, {
      slots: {
        default: "<div> TESTE </div>",
      },
      global: {
        plugins: [router],
      },
    });

    router.push("/contato");
    await router.isReady();

    await nextTick();
    await flushPromises();
    jest.advanceTimersByTime(1500);

    expect(beforeEachSpy).toHaveBeenCalled();
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("should animation must be undefined if slot is not passed", async () => {
    jest.useFakeTimers();

    const wrapper = mount(PageLoaderAnimation, {
      global: {
        plugins: [router],
      },
    });

    await nextTick();
    await flushPromises();

    expect(wrapper.html()).toMatchSnapshot();
  });
});
