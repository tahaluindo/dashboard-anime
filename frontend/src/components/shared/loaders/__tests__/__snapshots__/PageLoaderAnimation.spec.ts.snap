// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageLoaderAnimation test suit renders the default componenent 1`] = `
"<div class=\\"
      fixed
      inset-0
      z-0
      bg-transparent
      flex flex-col
      justify-center
      items-center
    \\" style=\\"\\"><span><div> TESTE </div></span><span class=\\"w-3 h-3 rounded-full bg-primary\\" style=\\"transform: translateX(-49.999999999999986px) scaleX(1) ;\\"></span></div>"
`;

exports[`PageLoaderAnimation test suit should animation must be undefined if slot is not passed 1`] = `
"<div class=\\"
      fixed
      inset-0
      z-0
      bg-transparent
      flex flex-col
      justify-center
      items-center
    \\" style=\\"display: none;\\"><span></span><span class=\\"w-3 h-3 rounded-full bg-primary\\"></span></div>"
`;
